src/main/java/com/puppycrawl/tools/checkstyle/checks/design/FinalClassCheck.java

```diff
@@ -253,14 +255,33 @@ public void leaveToken(DetailAST ast) {
      * @return true if given class should be declared as final otherwise false
      */
     private static boolean shouldBeDeclaredAsFinal(ClassDesc desc) {
-        return desc.isWithPrivateCtor()
+        return (desc.isWithPrivateCtor() || isPrivateWithDefaultCtor(desc))
```

Comment:
Following on the other comment, the way this code is written, your saying the ctor is not private but this is misleading since it is talking about a defined ctor.

It also seems odd we have `desc.isWithPrivateCtor() && !desc.isWithNonPrivateCtor()`. It seems like it is saying `true && !false`.

We may have to rename methods/fields to further clarify we are talking about defined ctors.

Reply:
The logic is really looks odd if class has private constructor then `desc.isWithPrivateCtor()` become true and if any other modifier is their `desc.isWithNonPrivateCtor()`  become false

@rnveach if the code is like 
```
class test5  
{
    private test5() {}  // ---> `desc.isWithPrivateCtor() == true`
     test5(int i) {}        //  ----->  `desc.isWithNonPrivateCtor() == false`
}
```
if totally remove `desc.isWithNonPrivateCtor()` this type of condition would be failed